### PR TITLE
Multiprocessing to the script that fetches receipt images

### PR DESCRIPTION
**Pull request by mtrovo**
Follow up to #35 adding support for download images in parallel and executing the download outside the constructor of the class FetchReceipts.

### CODE REVIEW

**Comment by cuducos** [comment]
Hi, @mtrovo, thanks for this PR.

In general it looks good, but trying to merge it to `master` not to `ec-fecth-reciepts` actually makes it difficult to follow up.

What I mean is that this code here is based on my first commit on PR #35, which was considerably improved after your code review on that very PR. Multiprocessing is a good idea (specially keeping the `limit` argument, as you did), but separating the constructor (as I did following your clever suggestion) also is a good idea. 

So my question actually is: how do we merge both things before bringing them to `master`?

My suggestion: would you mind closing this PR and  pulling the changes you made to `ec-fetch-receipts` (PR #35) and them we merge from there to `master`?

**Comment by cuducos** [comment]
Actually you could just change this PR https://github.com/blog/2224-change-the-base-branch-of-a-pull-request

**Comment by cuducos** [comment]
Actually… i just found out that I could do it myself. Gonna merge soon ; ) Tks

**Comment by cuducos** [comment]
> This branch has conflicts that must be resolved

@mtrovo can you resolve these conflicts so I can merge it?

File: src/fetch_receipts.py [hunk 1]
```diff
@@ -54,12 +57,36 @@ def __init__(self, target, limit=None):
...
-            else:
-                break
 
+    def run(self):
+        receipts_gen = (r for r in self.receipts if not os.path.exists(r.path))
+
+        with multiprocessing.Pool(processes=4) as pool:
+            while True:
+                if self.limit > 0:
+                    to_download = islice(receipts_gen, self.limit - self.progress['count'])
```
Comment (cuducos) [inline src/fetch_receipts.py]: `itertools.islice()` is new for me — and it's awesome! Thanks for that!

File: src/fetch_receipts.py [hunk 2]
```diff
@@ -54,12 +57,36 @@ def __init__(self, target, limit=None):
...
-                self.save(receipt)
-            else:
-                break
 
+    def run(self):
+        receipts_gen = (r for r in self.receipts if not os.path.exists(r.path))
+
+        with multiprocessing.Pool(processes=4) as pool:
+            while True:
+                if self.limit > 0:
```
Comment (cuducos) [inline src/fetch_receipts.py]: `if self.limit` already does the job. (No need for the `> 0` I guess).

**Comment by cuducos** [comment]
I'm gonna merge #35 to `master` and when you're able to sort out these conflicts multiprocessing is more than welcomed there ; )

**Comment by mtrovo** [comment]
@cuducos I made a change before your last comment actually, could you take a look?
I think it's good to merge.

**Comment by cuducos** [comment]
Ow… sorry, my bad! I missed that. Gonna take a look and merge soon if everything is fine.

File: src/fetch_receipts.py [hunk 3]
```diff
@@ -9,8 +9,14 @@
 import numpy as np
 import pandas as pd
 
+from itertools import islice
+from multiprocessing import Pool
 
-class Receipts:
+
+class Receipts(object):
```
Comment (cuducos) [inline src/fetch_receipts.py]: AFAIK there's no need to inherit from `object` in Python 3 — is there a specific reason to do this here?

File: src/fetch_receipts.py [hunk 4]
```diff
@@ -41,28 +47,31 @@ def all(self):
         }
         for dataset in self.datasets():
             data = pd.read_csv(dataset, parse_dates=[16], dtype=dtype)
-            yield from(Receipt(row) for row in data.itertuples())
+            yield from(Receipt(row, self.target) for row in data.itertuples() if not pd.isnull(row.document_id))
```
Comment (cuducos) [inline src/fetch_receipts.py]: If we keep `target` here, you're changing the output of this class without updating the documentation (__docstring__). Can you update that, please?

File: src/fetch_receipts.py [hunk 5]
```diff
@@ -41,28 +47,31 @@ def all(self):
...
 
 
 class Receipt:
 
-    def __init__(self, receipt):
+    def __init__(self, receipt, target):
```
Comment (cuducos) [inline src/fetch_receipts.py]: If we keep `target` here, you're changing the input of this class without updating the documentation (__docstring__). Can you update that, please?

File: src/fetch_receipts.py [hunk 6]
```diff
@@ -105,28 +114,49 @@ def run(target, limit=None):
...
+                    len(progress['errors'])
+                )
+                print(msg, end='\r')
+
+            # exit condition
+            if not limit or progress['count'] >= limit:
+                break
+    return progress
+
+def download_receipt(receipt):
```
Comment (cuducos) [inline src/fetch_receipts.py]: Would you mind adding documentation (`__docstring__`)?

File: src/fetch_receipts.py [hunk 7]
```diff
@@ -105,28 +114,49 @@ def run(target, limit=None):
...
+    if not os.path.exists(path):
+        try:
+            file_name, header = urlretrieve(receipt.url, path)
+            return ('ok', receipt, header)
+        except HTTPError as e:
+            return ('error', receipt, repr(e))
+    else:
+        return ('skipped', receipt, receipt.url)
+
+def print_report(progress):
```
Comment (cuducos) [inline src/fetch_receipts.py]: Would you mind adding documentation (`__docstring__`)?

File: src/fetch_receipts.py [hunk 8]
```diff
@@ -174,4 +204,5 @@ def run(target, limit=None):
     args = parser.parse_args()
 
     # run
-    run(args.target, args.limit)
+    result = run(args.target, args.limit)
+    print_report(result)
```
Comment (cuducos) [inline src/fetch_receipts.py]: What about `run()` calling `print_report()` at the end of its execution? Or is it an issue for multiprocessing (knowing when all the processes are done)?

Comment (lscasanova) [inline src/fetch_receipts.py]: I tried putting the `print_report()` inside `run()`, but then the script hangs after downloading the files.

Comment (cuducos) [inline src/fetch_receipts.py]: So we might just leave it as @mtrovo suggested ; ) Sorry guys

Comment (lscasanova) [inline src/fetch_receipts.py]: 👍

**Comment by cuducos** [comment]
@mtrovo and @lscasanova: many thanks for your contribution. I just pulled your commits to [`ec-fetch-receipts`](https://github.com/datasciencebr/serenata-de-amor/tree/ec-fetch-receipts), made some minor adjustments and I'm merging it to master.

**Pull request merged**